Call before/after_successful_authorization callbacks on create as well as new

mdeutsch opened the pull request:
### Summary

The before/after_successful_authorization callbacks added in #1064 are only called on the `new` action. I think they should also be called on `create` to cover situations where the user must explicitly authorize the client.

### Other Information

I also found that the callbacks can be run when authorization is not successful. I added some more tests and modified the existing code a bit.

This PR has some conflicts with my other PR, #1072. I'll resolve them but wanted to put this up for review.

cc @nattfodd

nbulaj commented:
So the callback will be invoked twice? Looks strange from my point of view...

Moreover, maybe it will be better to move the hooks under the Authorization class instead of controller? It will DRY the code at least

nattfodd commented:
> So the callback will be invoked twice? Looks strange from my point of view...

+1. I would rather rename callback in `create`(or both?)  into something different, so it would be 2 different callbacks.

> Moreover, maybe it will be better to move the hooks under the Authorization class instead of controller? It will DRY the code at least

As I mentioned in previous PR, we need controller's scope in that callback (i.e. for interacting with cookies/sessions).

nbulaj commented:
We can use existing callbacks, but add an extra argument - action ('new' or 'create'). Controller ais already passed with the context argument, so you already has an access to cookies or request.

mdeutsch commented:
I've rebased onto master and extracted some of the duplicated code. Now the callbacks are closer to the Authorization class and I think the intent is clearer: that the callbacks are invoked when the request is successfully authorized.

nbulaj commented on app/controllers/doorkeeper/authorizations_controller.rb:
```diff
@@ -84,7 +80,13 @@ def strategy
     def authorize_response
-      @authorize_response ||= strategy.authorize
+      @authorize_response ||= begin
+        valid = pre_auth.authorizable?
```
Let's rename it to `authorizable` to be closer to implementation and the essentials of this method. What do you think?

mdeutsch replied:
Good call -- I'll change it.

mdeutsch commented:
OK, I renamed the variable, added a changelog entry, and squashed/rebased.

nbulaj commented:
Got it, thank you for your work! :+1:
